webrender/src/clip_scroll_tree.rs

```diff
@@ -417,17 +417,18 @@ impl ClipScrollTree {
 
             node.push_gpu_node_data(gpu_node_data);
 
-            if node.children.is_empty() || node.combined_clip_outer_bounds.is_empty() {
```

Comment:
so why can we not check for the outer bounds here?

Reply:
We need to keep recursing even in this case in order to properly set the GPU node data for all child nodes (otherwise it might be stale). Additionally, once we support non-hierarchical clip chains, this optimization has to be removed or reworked.

Reply:
Aren't child nodes going to be invisible anyway if the outer bounds are empty?

Reply:
Yes. The issue here is when they were previously marked as visible, we need to now recurse to mark them as invisible. If we do not, the other bits of WebRender will believe they are still visible, but they won't have any corresponding GPU data. 

I think there's still a lot of adjustment we can do here. For instance, as I previously mentioned, we should further decouple the idea of a clip node and a clip chain in order to support complicated clipping situations. I'm working on this now. Additionally, the per-node GPU data is only holding transform information at the moment, so it makes sense that we only push this for nodes that adjust the transform (positioning nodes, but not clipping nodes).

These two things, along with allowing us to fix some Gecko issues, will also allow us to one day separate the clipping and scrolling/positioning trees completely.